scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+    )
+  )
+);
+```
+
+notes:
+We can do a similar thing for our program effect, adding our catchTag functionality in the pipe method after the generator.
+
+---
+
+```ts
+Effect.tryPromise({
+      try: () => fetch(`https://pokeapi.co/api/v2/pokemon/${id}`),
+      catch: () => new FetchError(),
+    }),
+```
+
+notes:
+Our pokemon app is hard coded to fetch data through a fetch call to a specific api, but what if we wanted to get the data from a different api, or a database, or maybe even add a cache layer within our application to check before making an api call. To do this traditionally you would have to refactor your code on every change, 
+
+```ts
+const getPokemon = (id: number, client: PokemonClient) =>
+  pipe(
+    client.pokemon.getById(id),
+    // ...
+  )
+```
+
+but what if we could abstract the signature of the data fetching to a dependency that we pass to our effects and could change the implementation of somewhere higher up the chain
+
+---
+
+### Adding Requirements
+(dependency injection)
+
+notes:
+Effect provides this behavior through the Requirements field of the Effect type which we have ignored up until now. Lets look at a simple example
+
+---
+```ts
+type Random = {
+  readonly next: () => Effect.Effect<never, never, number>
+}
+```
+
+notes:
+first we'll define the type signature of our dependency. Here it will be an object will one method called next that returns a Effect never never number
+
+---
+```ts
+import { Effect, Context } from "effect"
+
+type Random = {
+  readonly next: () => Effect.Effect<never, never, number>
+}
+
+const Random = Context.Tag<Random>()
+```
+
+notes:
+Next we'll import the context namespace from effect and use the tag function passing the type we just created as a generic. A tag is a unique identifier that represents a dependency of the type passed as the generic.
+
+---
+```ts
+// type: Effect<Random, never, void>
+const program = pipe(
+  Random,
+  // "random" type: Random
+  Effect.flatMap((random) => random.next()),
+  Effect.flatMap((randomNumber) => Effect.log(`random number: ${randomNumber}`))
+)
+```
+
+notes:
+We can use the tag just like any other effect, here piping it through flatMap. Within the flatmap function we get access to the value of type Random which we can call the next method on. Notice how the Random type now appears in the requirements field of the resulting Effect.
+
+---
+
+```ts
+// type: Effect<Random, never, void>
+const program = pipe(
+  Random,
+  // ...
+)
+
+Effect.runSync(program) // ERROR! Type 'Random' is not assignable to type 'never'
+```
+
+notes:
+If we try to run a program with a requirements field that is not 'never' we get a type error. Effect is ensuring we provide any dependencies required for our program before it can be run.
+
+---
+```ts
+// type: Effect<never, never, number>
+const runnable = program.pipe(
+  Effect.provideService(
+    Random,
+    Random.of({ next: () => Effect.succeed(Math.random()) })
```

Comment:
```ts
Effect.succeed(Math.random())
```

is always a mistake, executing side-effects eagerly should never be done/shown, it is important to use:

```ts
Random.of({ next: () => Effect.sync(() => Math.random()) })
```

otherwise `Effect.repeatN(500)(next())` will generate 500 times the same number.